c/common/adapters/httpapi_wince.cpp

```diff
@@ -0,0 +1,726 @@
+
+// data structure that contains all the information required to manage a connection
+class HTTP_HANDLE_DATA
+{
+public:
+    char   host[MAX_HOSTNAME];
+    SOCKET _sock_fd;
+    struct sockaddr_in _remoteHost;
+};
+
+// Thanks to Luca Calligaris for his article:
+// https://lcalligaris.wordpress.com/2011/04/07/implementing-a-secure-socket/
+
+#define SSL_OPT_NO_CHECK  0
+#define SSL_OPT_ISSUER_CHECK 1
+#define SSL_OPT_STRICT_CHECK 2
+
+#define SSL_OPT_DEFAULT   SSL_OPT_STRICT_CHECK
+
+static int g_SslOptions = SSL_OPT_DEFAULT;
+
+// callback called every time a certificate needs to be validated, depending on g_SslOptions global var it can perform
+// a strict validation (default) or just validate any certificate (may be needed for testing and debugging, not in production)
+int CALLBACK SSLValidateCertHook( DWORD  dwType, LPVOID pvArg, 
+    DWORD  dwChainLen, LPBLOB pCertChain, DWORD dwFlags)         
+{
+	HCERTSTORE hCertStore = NULL;
+	PCCERT_CONTEXT pCertInStore = NULL, pCertPassed = NULL, pPrevCertContext = NULL;
+	const LPCTSTR SystemStore[] = { _T("CA"), _T("ROOT"), _T("MY") };
+	unsigned int i;
+
+	//The only thing we can handle
+	if (dwType != SSL_CERT_X509)
+		return SSL_ERR_CERT_UNKNOWN;
+
+	//Do not check the certificate at all
+	if (g_SslOptions == SSL_OPT_NO_CHECK)
+		return SSL_ERR_OKAY;
+
+	//Get a context struct from the certificate blob
+	pCertPassed = CertCreateCertificateContext(X509_ASN_ENCODING, pCertChain->pBlobData, pCertChain->cbSize);
+
+	if (!pCertPassed)
+		return SSL_ERR_FAILED;
+
+	//Check for a matching certificate in the stores
+	for (i = 0; i<(sizeof(SystemStore) / sizeof(SystemStore[0])); i++)
+	{
+		//open the store
+		hCertStore = CertOpenSystemStore((HCRYPTPROV)NULL, SystemStore[i]);
+
+		if (!hCertStore)
+			continue;
+
+		pPrevCertContext = NULL;
+
+		do
+		{
+			//Find a certificate from the same issuer
+			pCertInStore = CertFindCertificateInStore(hCertStore, X509_ASN_ENCODING,
+				0, CERT_FIND_ISSUER_NAME, &pCertPassed->pCertInfo->Issuer, pPrevCertContext);
+
+			if (pCertInStore)
+			{
+				//It's enough if we have a certificate from te same issuer
+				if (g_SslOptions == SSL_OPT_ISSUER_CHECK)
+				{
+					CertFreeCertificateContext(pCertInStore);
+					CertCloseStore(hCertStore, 0);
+					CertFreeCertificateContext(pCertPassed);
+					return SSL_ERR_OKAY;
+				}
+				else if (g_SslOptions == SSL_OPT_STRICT_CHECK)
+				{
+					//Compare (exact match) the certificate with the one we have found
+					if (CertCompareCertificate(X509_ASN_ENCODING, pCertInStore->pCertInfo, pCertPassed->pCertInfo))
+					{
+						CertFreeCertificateContext(pCertInStore);
+						CertCloseStore(hCertStore, 0);
+						CertFreeCertificateContext(pCertPassed);
+						return SSL_ERR_OKAY;
+					}
+				}
+			}
+
+			pPrevCertContext = pCertInStore;
+		} while (pCertInStore);
+
+		CertCloseStore(hCertStore, 0);
+	}
+
+	if (pCertInStore)
+		CertFreeCertificateContext(pCertInStore);
+
+	if (hCertStore)
```

Comment:
In line 112 the CertStore is already closed. There is no reason I believe to close it again.